### PR TITLE
IPMVAL-3740 fix

### PR DESCRIPTION
**Pull request by FrancoisRegisDegott-eaton**
* IPMVAL-3740: fix rule evaluation process on ADD rule
Signed-off-by: Francois-Regis Degott <FrancoisRegisDegott@eaton.com>

### CODE REVIEW

File: src/fty_alert_engine_server.cc [hunk 1]
```diff
@@ -417,7 +419,7 @@ static void get_rule(mlm_client_t* client, const char* name, AlertConfiguration&
     mtxAlertConfig.unlock();
 
     if (!found) {
-        log_debug("rule not found (%s)", name);
+        log_debug("rule not found (name: %s)", name);
```
Comment (ericclappier-eaton) [inline src/fty_alert_engine_server.cc:423]: Crash if name is null, better to test it at the beginning

File: src/fty_alert_engine_server.cc [hunk 2]
```diff
@@ -610,16 +635,23 @@ static void update_rule(mlm_client_t* client, const char* json_representation, c
...
 }
 
 static void delete_rules(mlm_client_t* client, RuleMatcher* matcher, AlertConfiguration& ac)
 {
     std::map<std::string, std::vector<PureAlert>> alertsToSend;
     std::vector<std::string>                      rulesDeleted;
+
     mtxAlertConfig.lock();
+    int rv = ac.deleteRules(matcher, alertsToSend, rulesDeleted);
+    mtxAlertConfig.unlock();
```
Comment (ericclappier-eaton) [inline src/fty_alert_engine_server.cc:652]: maybe an issue if send_alerts is calling below when another thread delete some elements

File: src/fty_alert_engine_server.cc [hunk 3]
```diff
@@ -1023,194 +1027,176 @@ void fty_alert_engine_stream(zsock_t* pipe, void* args)
 void fty_alert_engine_mailbox(zsock_t* pipe, void* args)
 {
     char* name = static_cast<char*>(args);
+    assert(name);
```
Comment (ericclappier-eaton) [inline src/fty_alert_engine_server.cc:1031]: return better

File: src/fty_alert_engine_server.cc [hunk 4]
```diff
@@ -857,164 +894,131 @@ void fty_alert_engine_stream(zsock_t* pipe, void* args)
...
-        if (zmessage) {
-            // Here we can have a message with arbitrary topic, but according protocol
-            // first frame must be one of the following:
-            //  * METRIC_UNAVAILABLE
-            char* command = zmsg_popstr(zmessage);
-            if (streq(command, "METRICUNAVAILABLE")) {
-                char* metrictopic = zmsg_popstr(zmessage);
-                if (metrictopic) {
-                    check_metrics(client, metrictopic, alertConfiguration);
+        if (which == mlm_client_msgpipe(client)) {
```
Comment (ericclappier-eaton) [inline src/fty_alert_engine_server.cc:981]: else if ??

File: src/fty_alert_engine_server.cc [hunk 5]
```diff
@@ -1023,194 +1027,176 @@ void fty_alert_engine_stream(zsock_t* pipe, void* args)
...
-        if (streq(mlm_client_subject(client), RULES_SUBJECT)) {
-            // According RFC we expect here a messages
-            // with the topic:
-            //   * RULES_SUBJECT
-            // Here we can have:
-            //  * request for list of rules
-            //  * get detailed info about the rule
-            //  * new/update rule
-            //  * touch rule
+        if (which == mlm_client_msgpipe(client)) {
```
Comment (ericclappier-eaton) [inline src/fty_alert_engine_server.cc:1107]: else if better

**Review by ericclappier-eaton — CHANGES_REQUESTED** [review]

File: src/autoconfig.cc [hunk 6]
```diff
@@ -156,56 +162,68 @@ void Autoconfig::main (zsock_t* pipe, char* name)
...
+            else {
+                log_debug("%s: command not handled (%s)", name, cmd);
+            }
 
             zstr_free(&cmd);
             zmsg_destroy(&msg);
             continue;
         }
 
+        //TODO rewrite that crappy message processing
```
Comment (aquette) [inline src/autoconfig.cc:225]: s/message processing/agent/
:D ;)

File: src/fty_alert_engine.cc [hunk 7]
```diff
@@ -110,35 +109,26 @@ int main(int argc, char** argv)
...
-    while (true) {
-        char* messageS = zstr_recv(ag_server_stream);
-        if (messageS) {
-            puts(messageS);
-            free(messageS);
-        } else {
-            log_info("interrupted");
-            break;
-        }
+    log_info("%s started", ENGINE_AGENT_NAME);
```
Comment (aquette) [inline src/fty_alert_engine.cc:112]: should be alert_engine_name (even if should be the same)

File: src/fty_alert_engine.cc [hunk 8]
```diff
@@ -110,35 +109,26 @@ int main(int argc, char** argv)
...
-            log_info("interrupted");
+    // main loop, accept any message back from server
+    // copy from src/malamute.c under MPL license
+    while (!zsys_interrupted) {
+        char* msg = zstr_recv(ag_server_mailbox);
+        if (!msg)
             break;
-        }
+
+        log_debug("%s: recv msg '%s'", ENGINE_AGENT_NAME, msg);
```
Comment (aquette) [inline src/fty_alert_engine.cc:121]: ditto above...

File: src/fty_alert_engine_server.cc (hunk 2 above)
Comment (aquette) [reply src/fty_alert_engine_server.cc:652]: to be mutexed so

**Review by aquette — APPROVED** [review]
LGTM as long as @eric3873 's comments are addressed!

File: src/fty_alert_engine_server.cc (hunk 2 above)
Comment (jeffdegott) [reply src/fty_alert_engine_server.cc:651]: reverted

Comment (FrancoisRegisDegott-eaton) [reply src/fty_alert_engine_server.cc:652]: viewed with Eric. All is ok  ;)

**Review by ericclappier-eaton — APPROVED** [review]

**Pull request merged**